### PR TITLE
Added an implementation to Stratify Data

### PR DESCRIPTION
**Pull request by Abilityguy**
Reference #2662 

@rcurtin challenge completed!

The current implementation seems to be faster than the one in the issue. 

Dataset 1: covertype dataset (https://www.mlpack.org/datasets/covertype-small.data.csv.gz)

```
Dataset size - 54 x 100000
Test ratio - 0.3

Previous time:
[INFO ]   loading_data: 0.458566s
[INFO ]   saving_data: 2.796008s
[INFO ]   total_time: 0.641220s

Improved time:
[INFO ]   loading_data: 0.366347s
[INFO ]   saving_data: 2.499058s
[INFO ]   total_time: 0.413254s
```

Dataset 2: MNIST train dataset from Kaggle (https://www.kaggle.com/c/digit-recognizer/data)
```
Dataset size - 784 x 42000
Test ratio - 0.2

Previous time:
[INFO ]   loading_data: 2.912084s
[INFO ]   saving_data: 19.503108s
[INFO ]   total_time: 3.998145s

Improved time:
[INFO ]   loading_data: 2.153167s
[INFO ]   saving_data: 14.734252s
[INFO ]   total_time: 2.406783s
```

I have added some comments to convey the basic idea in stratified_split_data.hpp
Let me know what you guys think.

### CODE REVIEW

File: src/mlpack/core/data/stratified_split_data.hpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+  arma::uvec Indexes;
+  Indexes.set_size(inputLabel.n_cols);
+
+  //Single index vector which adds train Indexes from the start and test
+  //indexes from the end
+  size_t trainIdx = 0;
+  size_t testIdx = inputLabel.n_cols - 1;
+  std::unordered_map<U, size_t> labelMap;
+
+  if (shuffleData)
```
Comment (zoq) [inline src/mlpack/core/data/stratified_split_data.hpp:106]: You could combine the two cases, by using:

```
arma::uvec order = arma::linspace<arma::uvec>(
        0, input.n_cols - 1, input.n_cols);
```

outside the if/else case, iterate over the order in each case and only shuffle the order if `shuffleData` is true. Does that makes sense?

Comment (Abilityguy) [reply src/mlpack/core/data/stratified_split_data.hpp:106]: Makes sense. I will make the changes.
Thanks for the review @zoq!.

**Comment by Abilityguy** [comment]
Hi @zoq. I have made the changes.

I was thinking if it would be better to have the stratified split templates in the file split_data.hpp file itself. 
This would help to refactor the code in preprocess_split_main.cpp into a single call to data::Split(..., stratify_data) with stratify_data being an extra flag that indicates whether to stratify or not.

What is your take on this?

File: src/mlpack/methods/preprocess/preprocess_split_main.cpp [hunk 2]
```diff
@@ -67,7 +68,19 @@ BINDING_EXAMPLE(
     "\n\n" +
     PRINT_CALL("preprocess_split", "input", "X", "input_labels", "y",
         "test_ratio", 0.3, "training", "X_train", "training_labels", "y_train",
-        "test", "X_test", "test_labels", "y_test"));
+        "test", "X_test", "test_labels", "y_test") +
+    "Also by default the dataset is not stratified; you can provide the " +
```
Comment (Yashwants19) [inline src/mlpack/methods/preprocess/preprocess_split_main.cpp:72]: If this example is not related with above example, then for improving the readability, we can split the example into two different macros(`BINDING_EXAMPLE()`).
For E.g.
```cpp
BINDING_EXAMPLE(
    "So, a simple...");
BINDING_EXAMPLE(
    /*Stratify_data Example*/);
```
Let me what do you think?

File: src/mlpack/methods/preprocess/preprocess_split_main.cpp [hunk 3]
```diff
@@ -148,22 +164,45 @@ static void mlpackMain()
...
+          if (IO::HasParam("training"))
+            IO::GetParam<arma::mat>("training") = std::move(get<0>(value));
+          if (IO::HasParam("test"))
+            IO::GetParam<arma::mat>("test") = std::move(get<1>(value));
+          if (IO::HasParam("training_labels"))
+            IO::GetParam<arma::Mat<size_t>>("training_labels") =
+                std::move(get<2>(value));
+          if (IO::HasParam("test_labels"))
+            IO::GetParam<arma::Mat<size_t>>("test_labels") =
+                std::move(get<3>(value));
```
Comment (Yashwants19) [inline src/mlpack/methods/preprocess/preprocess_split_main.cpp:185]:
```suggestion
        Log::Info << "Training data contains "
            << get<0>(value).n_cols << " points." << endl;
        Log::Info << "Test data contains "
            << get<1>(value).n_cols << " points." << endl;

        if (IO::HasParam("training"))
          IO::GetParam<arma::mat>("training") = std::move(get<0>(value));
        if (IO::HasParam("test"))
          IO::GetParam<arma::mat>("test") = std::move(get<1>(value));
        if (IO::HasParam("training_labels"))
          IO::GetParam<arma::Mat<size_t>>("training_labels") =
              std::move(get<2>(value));
        if (IO::HasParam("test_labels"))
          IO::GetParam<arma::Mat<size_t>>("test_labels") =
              std::move(get<3>(value));
```
I guess the indentation is incorrect.

File: src/mlpack/methods/preprocess/preprocess_split_main.cpp [hunk 4]
```diff
@@ -102,11 +116,13 @@ static void mlpackMain()
...
   const double testRatio = IO::GetParam<double>("test_ratio");
   const bool shuffleData = IO::GetParam<bool>("no_shuffle");
+  const bool stratifyData = IO::GetParam<bool>("stratify_data");
 
   if (IO::GetParam<int>("seed") == 0)
     mlpack::math::RandomSeed(std::time(NULL));
   else
-    mlpack::math::RandomSeed((size_t) IO::GetParam<int>("seed"));
+    mlpack::math::RandomSeed((size_t)
+     IO::GetParam<int>("seed"));
```
Comment (Yashwants19) [inline src/mlpack/methods/preprocess/preprocess_split_main.cpp:125]:
```suggestion
    mlpack::math::RandomSeed((size_t) IO::GetParam<int>("seed"));
```
I think so this was correct only, Or I am missing something.

**Review by Yashwants19 — COMMENTED** [review]
I guess we should also add some tests here, both for method(`tests/split_data_test.cpp`) as well as bindings(`main_tests/preprocess_split_test.cpp`).

File: src/mlpack/methods/preprocess/preprocess_split_main.cpp (hunk 2 above)
Comment (Abilityguy) [reply src/mlpack/methods/preprocess/preprocess_split_main.cpp:72]: You are right @Yashwants19. The examples are not related and it makes sense to split them up. Will make this change.

File: src/mlpack/methods/preprocess/preprocess_split_main.cpp (hunk 4 above)
Comment (Abilityguy) [reply src/mlpack/methods/preprocess/preprocess_split_main.cpp:125]: Nope your change makes sense. This doesn't cross the 80 character limit.

**Comment by Abilityguy** [comment]
Hey @Yashwants19 , thanks for the review. 
I have made changes according to your review. I added a separate binding example. 
I also deleted the stratified_split_data.hpp file and moved it's template to split_data.hpp. 
There seems to be a documentation fail. It might be due to a mistake by me while documenting. I am not sure about it's fix though. If someone can let me know how to fix it, I will make the change.

File: src/mlpack/core/data/split_data.hpp [hunk 5]
```diff
@@ -176,15 +294,24 @@ std::tuple<arma::Mat<T>, arma::Mat<T>, arma::Row<U>, arma::Row<U>>
 Split(const arma::Mat<T>& input,
       const arma::Row<U>& inputLabel,
       const double testRatio,
-      const bool shuffleData = true)
+      const bool shuffleData = true,
+      const bool stratifyData = false)
```
Comment (Yashwants19) [inline src/mlpack/core/data/split_data.hpp:298]: We can resolve the issue by adding documentation for  `stratifyData` here.

Comment (Abilityguy) [reply src/mlpack/core/data/split_data.hpp:300]: Thanks for pointing that out!. Made the fix.

File: src/mlpack/core/data/split_data.hpp [hunk 6]
```diff
@@ -17,6 +17,124 @@
...
+   * 1 -- train set ( 5*0.2 == 6*0.2 ) (After casting)
+   *
+   * Finally
+   * train set,
+   * 0 0 0 0 1 1 1 1 1 (4 0s, 5 1s)
+   *
+   * test set,
+   * 0 1
+   */
+  arma::uvec Indexes;
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp:99]: Let's use `indices` here, also keep in mind that mlpack parameter start with a lower case.

File: src/mlpack/core/data/split_data.hpp [hunk 7]
```diff
@@ -17,6 +17,124 @@
...
+    }
+    else
+    {
+      Indexes[trainIdx] = i;
+      trainIdx += 1;
+    }
+    labelMap[label] += 1;
+  }
+
+  labelMap.clear();
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp:131]: The map is going to be deconstructed anyway, so the clear looks redundant.

File: src/mlpack/core/data/split_data.hpp [hunk 8]
```diff
@@ -17,6 +17,124 @@
...
+   * 0 -- train set ( 2*0.2 == 3*0.2 ) (After casting)
+   * 0 -- train set ( 3*0.2 == 4*0.2 ) (After casting)
+   * 0 -- test set  ( 4*0.2 <  5*0.2 ) (After casting)
+   *
+   * 1 -- train set ( 0*0.2 == 1*0.2 ) (After casting)
+   * 1 -- train set ( 1*0.2 == 2*0.2 ) (After casting)
+   * 1 -- train set ( 2*0.2 == 3*0.2 ) (After casting)
+   * 1 -- train set ( 3*0.2 == 4*0.2 ) (After casting)
+   * 1 -- test set  ( 4*0.2 <  5*0.2 ) (After casting)
+   * 1 -- train set ( 5*0.2 == 6*0.2 ) (After casting)
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp:90]:
```suggestion
   * Whenever we encounter a label, we calculate
   * current_count * test_ratio --- labelMap[label] * testRatio and
   * current_count+1 * test_ratio --- (labelMap[label] + 1) * testRatio
   *
   * We then static_cast these counts to size_t to remove their decimal points.
   * If in this case, our integer counts are same then we add to our train set.
   * If there is a difference in counts, then we add to our test set
   *
   * Considering our example
   * 0 -- train set ( 0 * 0.2 == 1 * 0.2 ) (After casting)
   * 0 -- train set ( 1 * 0.2 == 2 * 0.2 ) (After casting)
   * 0 -- train set ( 2 * 0.2 == 3 * 0.2 ) (After casting)
   * 0 -- train set ( 3 * 0.2 == 4 * 0.2 ) (After casting)
   * 0 -- test set  ( 4 * 0.2 <  5 * 0.2 ) (After casting)
   *
   * 1 -- train set ( 0 * 0.2 == 1 * 0.2 ) (After casting)
   * 1 -- train set ( 1 * 0.2 == 2 * 0.2 ) (After casting)
   * 1 -- train set ( 2 * 0.2 == 3 * 0.2 ) (After casting)
   * 1 -- train set ( 3 * 0.2 == 4 * 0.2 ) (After casting)
   * 1 -- test set  ( 4 * 0.2 <  5 * 0.2 ) (After casting)
   * 1 -- train set ( 5 * 0.2 == 6 * 0.2 ) (After casting)
```

Just to be consistent with the rest of the codebase.

**Review by zoq — COMMENTED** [review]
I think the plan right now is to add some tests for the new feature?

**Comment by Abilityguy** [comment]
Yes, I will look into adding a few tests.

File: src/mlpack/core/data/split_data.hpp [hunk 9]
```diff
@@ -17,6 +17,123 @@
...
+      indices[trainIdx] = i;
+      trainIdx += 1;
+    }
+    labelMap[label] += 1;
+  }
+
+  testData = input.cols(indices.subvec(trainIdx, indices.n_rows-1));
+  testLabel = inputLabel.cols(indices.subvec(trainIdx, indices.n_rows-1));
+  trainData = input.cols(indices.subvec(0, trainIdx-1));
+  trainLabel = inputLabel.cols(indices.subvec(0, trainIdx-1));
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp:134]:
```suggestion
  testData = input.cols(indices.subvec(trainIdx, indices.n_rows - 1));
  testLabel = inputLabel.cols(indices.subvec(trainIdx, indices.n_rows - 1));
  trainData = input.cols(indices.subvec(0, trainIdx - 1));
  trainLabel = inputLabel.cols(indices.subvec(0, trainIdx - 1));
```

File: src/mlpack/tests/split_data_test.cpp [hunk 10]
```diff
@@ -212,3 +212,114 @@ TEST_CASE("SplitLabeledDataLargerTest", "[SplitDataTest]")
 
   CheckDuplication(std::get<2>(value), std::get<3>(value));
 }
+
+/**
+ * Check that 0 test ratio results in a full train set for stratified split
```
Comment (zoq) [inline src/mlpack/tests/split_data_test.cpp:217]:
```suggestion
 * Check that 0 test ratio results in a full train set for stratified split.
```

Missing stop at the end.

File: src/mlpack/tests/split_data_test.cpp [hunk 11]
```diff
@@ -212,3 +212,114 @@ TEST_CASE("SplitLabeledDataLargerTest", "[SplitDataTest]")
...
+
+  const auto value = Split(input, labels, test_ratio, false, true);
+  REQUIRE(std::get<0>(value).n_cols == 15);
+  REQUIRE(std::get<1>(value).n_cols == 0);
+  REQUIRE(std::get<2>(value).n_cols == 15);
+  REQUIRE(std::get<3>(value).n_cols == 0);
+}
+
+/**
+ * Check that 1 test ratio results in a full test set for stratified split
```
Comment (zoq) [inline src/mlpack/tests/split_data_test.cpp:236]:
```suggestion
 * Check that 1 test ratio results in a full test set for stratified split.
```

Missing stop at the end.

File: src/mlpack/tests/split_data_test.cpp [hunk 12]
```diff
@@ -212,3 +212,114 @@ TEST_CASE("SplitLabeledDataLargerTest", "[SplitDataTest]")
...
+
+  const auto value = Split(input, labels, test_ratio, false, true);
+  REQUIRE(std::get<0>(value).n_cols == 0);
+  REQUIRE(std::get<1>(value).n_cols == 15);
+  REQUIRE(std::get<2>(value).n_cols == 0);
+  REQUIRE(std::get<3>(value).n_cols == 15);
+}
+
+/**
+ * Check if data is stratified according to labels
```
Comment (zoq) [inline src/mlpack/tests/split_data_test.cpp:255]:
```suggestion
 * Check if data is stratified according to labels.
```

Missing stop at the end.

File: src/mlpack/core/data/split_data.hpp [hunk 13]
```diff
@@ -17,6 +17,123 @@
...
+  if (shuffleData)
+  {
+    order = arma::shuffle(order);
+  }
+
+  for (auto i: order)
+  {
+    auto label = inputLabel[i];
+    if (static_cast<size_t>(labelMap[label]*testRatio) <
+        static_cast<size_t>((labelMap[label]+1)*testRatio))
```
[the remark this answers is not in the document]
Comment (Abilityguy) [reply src/mlpack/core/data/split_data.hpp:118]: The casting is to actually mimic the floor operation. Since testRatio is of type double, the calculation results in a double output. I wanted to remove the decimal points and just keep the integer part of this double and hence casted it to size_t. Will add space between the ```*, +```.

File: src/mlpack/core/data/split_data.hpp [hunk 14]
```diff
@@ -14,9 +14,139 @@
...
+      indices[trainIdx] = i;
+      trainIdx += 1;
+    }
+    else
+    {
+      indices[testIdx] = i;
+      testIdx -= 1;
+    }
+    labelMap[label] += 1;
+  }
```
Comment (rcurtin) [inline src/mlpack/core/data/split_data.hpp:129]: When I said it could be done in a single pass over labels and input matrix, I think I misspoke---I believe that this first pass over the labels is necessary in order to compute the size of the training data and labels, but then in a second pass over the labels and a first pass over the data you can then copy everything to the output matrices efficiently. :+1: (See my comments below for more on that.)

File: src/mlpack/core/data/split_data.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (Abilityguy) [reply src/mlpack/core/data/split_data.hpp]: Replaced the static_cast with the floor function so the calculation makes more sense now.

File: src/mlpack/core/data/split_data.hpp [hunk 15]
```diff
@@ -14,9 +14,139 @@
...
+
+  for (auto i: order)
+  {
+    auto label = inputLabel[i];
+
+    if (floor(labelMap[label] * testRatio) ==
+        floor((labelMap[label] + 1) * testRatio))
+    {
+      indices[trainIdx] = i;
```
Comment (rcurtin) [inline src/mlpack/core/data/split_data.hpp:120]: I think it would be possible to avoid the use of `indices`, if you wanted.  I believe that you can, in the first pass, simply compute the number of training points for each class.  (So instead of `indices`, you'd hold something like `arma::uvec trainingClassCounts`, where at the end of the loop `trainingClassCounts[i]` is the number of training points that are needed from class `i`.)

Then, in the second pass when you extract the data from `input` and `inputLabel` into `trainData` and `trainLabel`, you would also maintain something like `arma::uvec currentClassCounts`, and each time you encounter a point with class `i` where `currentClassCounts[i] <= trainingClassCounts[i]`, you copy the label and the point to the training set and increment `currentClassCounts[i]`.  Otherwise, you copy the label and point to the test set.

Let me know what you think---it's just an idea. :)  It could save a little bit of memory.

File: src/mlpack/core/data/split_data.hpp [hunk 16]
```diff
@@ -14,9 +14,139 @@
...
+  if (trainSize > 0)
+  {
+    trainData = input.cols(indices.subvec(0, trainIdx - 1));
+    trainLabel = inputLabel.cols(indices.subvec(0, trainIdx - 1));
+  }
+  if (trainSize < input.n_cols)
+  {
+    testData = input.cols(indices.subvec(trainIdx, indices.n_rows - 1));
+    testLabel = inputLabel.cols(indices.subvec(trainIdx, indices.n_rows - 1));
+  }
```
Comment (rcurtin) [inline src/mlpack/core/data/split_data.hpp:147]: Actually this ends up taking four passes over the data! :smile:  This is because each call to `cols()` to extract a new matrix will take an entire pass over `input`.

If you wanted to make it more efficient, you could write a single loop over `input` that copies each column at a time.  However, ideally we want to pass over `input` and `inputLabel` contiguously (i.e. access the first column, then the second column, etc...) for the sake of speed.  So in order to do that, the strategy I suggested above to avoid using `indices` might be necessary.

But, in any case, you could time both ways.  I'm pretty sure that the strategy I proposed could be faster but I'm not absolutely certain. :+1:

File: src/mlpack/methods/preprocess/preprocess_split_main.cpp [hunk 17]
```diff
@@ -69,6 +69,21 @@ BINDING_EXAMPLE(
         "test_ratio", 0.3, "training", "X_train", "training_labels", "y_train",
         "test", "X_test", "test_labels", "y_test"));
 
+BINDING_EXAMPLE(
+    "There is an option to stratify the dataset according to the labels." +
+    PRINT_PARAM_STRING("stratify_data") + " option to stratify the data;"
+    "an example to stratify the data is" +
```
Comment (rcurtin) [inline src/mlpack/methods/preprocess/preprocess_split_main.cpp:75]: These sentences seem really redundant.  Maybe something more like:

"To use stratified sampling so that the ratio of classes in the training and test set match the ratio of classes in the input data, we can specify the " + PRINT_PARAM_STRING("stratify_data") + " option."

(Or something like this.)

File: src/mlpack/methods/preprocess/preprocess_split_main.cpp [hunk 18]
```diff
@@ -125,6 +142,7 @@ static void mlpackMain()
   {
     ReportIgnoredParam({{ "input_labels", true }}, "training_labels");
     ReportIgnoredParam({{ "input_labels", true }}, "test_labels");
+    ReportIgnoredParam("stratify_data", "input_labels is not provided");
```
Comment (rcurtin) [inline src/mlpack/methods/preprocess/preprocess_split_main.cpp:145]: I don't think that this works as intended.  Can you check the documentation for `ReportIgnoredParam`?

File: src/mlpack/tests/split_data_test.cpp [hunk 19]
```diff
@@ -212,3 +212,113 @@ TEST_CASE("SplitLabeledDataLargerTest", "[SplitDataTest]")
...
+
+/**
+ * Check if data is stratified according to labels.
+ */
+TEST_CASE("StratifiedSplitDataResultTest", "[SplitDataTest]")
+{
+  mat input(5, 24);
+  input.randu();
+
+  // Set the lables to 4 0s, 8 1s and 12 2s.
```
Comment (rcurtin) [inline src/mlpack/tests/split_data_test.cpp:262]:
```suggestion
  // Set the labels to 4 0s, 8 1s and 12 2s.
```
:)

File: src/mlpack/tests/split_data_test.cpp [hunk 20]
```diff
@@ -212,3 +212,113 @@ TEST_CASE("SplitLabeledDataLargerTest", "[SplitDataTest]")
...
+
+/**
+ * Check that test ratio of 1 results in a full test set for stratified split.
+ */
+TEST_CASE("TotalRatioStratifiedSplitData", "[SplitDataTest]")
+{
+  mat input(2, 15);
+  input.randu();
+
+  // Set the lables to 5 0s and 10 1s.
```
Comment (rcurtin) [inline src/mlpack/tests/split_data_test.cpp:243]:
```suggestion
  // Set the labels to 5 0s and 10 1s.
```
:)

File: src/mlpack/tests/split_data_test.cpp [hunk 21]
```diff
@@ -212,3 +212,113 @@ TEST_CASE("SplitLabeledDataLargerTest", "[SplitDataTest]")
...
+
+/**
+ * Check that test ratio of 0 results in a full train set for stratified split.
+ */
+TEST_CASE("ZeroRatioStratifiedSplitData", "[SplitDataTest]")
+{
+  mat input(2, 15);
+  input.randu();
+
+  // Set the lables to 5 0s and 10 1s.
```
Comment (rcurtin) [inline src/mlpack/tests/split_data_test.cpp:224]:
```suggestion
  // Set the labels to 5 0s and 10 1s.
```

File: src/mlpack/tests/split_data_test.cpp [hunk 22]
```diff
@@ -212,3 +212,113 @@ TEST_CASE("SplitLabeledDataLargerTest", "[SplitDataTest]")
...
+  Row<size_t> three_label(32);
+
+  zero_label.fill(0);
+  one_label.fill(1);
+  two_label.fill(2);
+  three_label.fill(3);
+
+  Row<size_t> labels = arma::join_rows(zero_label, one_label);
+  labels = arma::join_rows(labels,two_label);
+  labels = arma::join_rows(labels,three_label);
```
Comment (rcurtin) [inline src/mlpack/tests/split_data_test.cpp:308]:
```suggestion
  labels = arma::join_rows(labels, two_label);
  labels = arma::join_rows(labels, three_label);
```
Just a simple style fix. :)

**Review by rcurtin — COMMENTED** [review]
Hey @abilityguy, thanks for putting the time into this!  It looks good.

I have some questions about your benchmarks.  Can you tell us about how you ran them and how you compiled them?  The `saving_time` and `loading_time` timers seem to be *longer* than the `total_time`, which is unexpected.  Also, `saving_time` and `loading_time` are just the times for `data::Load()` and `data::Save()`, so I would not expect them to be different between runs.

If you wanted to benchmark your stratification function specifically, you could add

```
Timer::Start("stratify_data");
```

and

```
Timer::Stop("stratify_data");
```

on either side of the call to `data::Split()` in `preprocess_split_main.cpp`.  (I guess it wouldn't hurt to have that timer be a part of the program, but it's up to you whether you just want to make that as a local change for benchmarking or add it to the program.)

Hopefully my review comments are helpful!  The algorithm you have used is quite clever (it took me a little while to understand it fully) but I think I might have suggested some things that could improve it further.  Let me know what you think!  :+1:

File: src/mlpack/core/data/split_data.hpp [hunk 23]
```diff
@@ -14,9 +14,139 @@
...
+   * Basic idea:
+   * Let us say we have to stratify a dataset based on labels:
+   * 0 0 0 0 0 1 1 1 1 1 1 (5 0s, 6 1s)
+   *
+   * Let our test ratio be 0.2:
+   * We visit each label and keep the count of each label in our unordered map
+   *
+   * Whenever we encounter a label, we calculate the floored values of
+   * current_count * test_ratio --- labelMap[label] * testRatio and
+   * current_count+1 * test_ratio --- (labelMap[label] + 1) * testRatio
```
[the remark this answers is not in the document]
Comment (Abilityguy) [reply src/mlpack/core/data/split_data.hpp:73]: ```---``` was just meant to point to the corresponding code that calculates this in the code. I can see how this is ambiguous and will change it.

File: src/mlpack/core/data/split_data.hpp (hunk 16 above)
Comment (Abilityguy) [reply src/mlpack/core/data/split_data.hpp:147]: Completely slipped my mind about the passes in cols!. 😆 
Thanks for pointing that out. 

Removing the ```indices``` vector makes sense. Will experiment and get back on this.

File: src/mlpack/methods/preprocess/preprocess_split_main.cpp (hunk 18 above)
Comment (Abilityguy) [reply src/mlpack/methods/preprocess/preprocess_split_main.cpp:134]: What I wanted to do here was to provide the user a warning if he provides the ```stratify_data``` parameter without providing the ```input_labels``` file. 
In this case, the ```stratify_data``` parameter is ignored.
Since no labels are passed to stratify from, we go ahead and perform a normal split. 

I went through the ```mlpack::util``` docs and the ```ReportIgnoredParam``` seemed like the function to use.
According to the ```ReportIgnoredParam``` docs, 
```If the given parameter is passed, report that it is ignored, supplying a custom reason.```

In this case, the ```stratify_data``` parameter is ignored because the ```input_labels``` parameter is not provided. 

Since this call is made inside the ```else``` block where the ```if``` checks if the ```input_labels``` file is provided, I thought this would be the right place to include it.

Let me know if this was not the right way to do it.

**Comment by rcurtin** [comment]
Interesting, so it was slower when you used `labelMap` and `testLabelMap` instead of `indices`?  Perhaps I was wrong---I would have expected the map approach to be quicker, but intuition is wrong something. :smile:

**Comment by Abilityguy** [comment]
Yup. I too expected the opposite. 
I made another check and recorded the timing over 5 runs and got similar results. The ```indices``` approach is faster compared to the ```labelMap``` and ```testLabelMap``` approach.
Of course, the ```indices``` approach takes more memory so it's a trade off.

**Comment by zoq** [comment]
> Yup. I too expected the opposite.
> I made another check and recorded the timing over 5 runs and got similar results. The `indices` approach is faster compared to the `labelMap` and `testLabelMap` approach.
> Of course, the `indices` approach takes more memory so it's a trade off.
> Let me know which implementation would be better.

The timings are pretty close, maybe the hashing isn't perfect and the lookup is not O(1)? Personally I would go with the current approach just to save some memory.

**Comment by Abilityguy** [comment]
Hey @zoq, it seems like you are right. The hashing may not have been ```O(1)```. I made another implementation where I replaced the ```unordered_map``` with an ```arma::uvec``` vector. 

I had to make an extra pass to figure out the maximum value of the label. (```inputLabel.max()```). But I found that this implementation achieves the best of both worlds by saving memory and being faster.

The times are listed below (average of 5).

```
Dataset 1: MNIST dataset

StratifiedSplit (with indicies vector):  0.123578 s
StratifiedSplit (with unordered map):    0.125799 s
StratifiedSplit (with arma::uvec):       0.1226292 s
```

```
Dataset 2: Covertype dataset

StratifiedSplit (with indicies vector):  0.0388812 s
StratifiedSplit (with unordered map):    0.040097 s
StratifiedSplit (with arma::uvec):       0.0356198 s
```

File: src/mlpack/core/data/split_data.hpp [hunk 24]
```diff
@@ -17,6 +17,136 @@
...
+   */
+  size_t trainIdx = 0;
+  size_t testIdx = 0;
+  size_t trainSize = 0;
+  size_t testSize = 0;
+  arma::uvec labelMap;
+  arma::uvec testLabelMap;
+  U maxLabel = inputLabel.max();
+
+  labelMap.zeros(maxLabel+1);
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp:103]: Hm, so this could be inefficient if I have a dataset with labels 0, 1000000, only? Also, this doesn't really work for floating point labels.

Comment (Abilityguy) [reply src/mlpack/core/data/split_data.hpp:103]: Yes, it would be inefficient if there is a dataset with 1000000 0s only. 
But in this case, the user can opt for a regular split instead of a stratified split. 

This method is a method to stratify discrete values only. For continuous label stratification that is, floating point labels this method won't work well. 
I didn't find a general method for continuous label stratification that everyone uses so I am not sure what we could go about implementing. There are two approaches that are possible. 

* One method the user can do is to bin the floating point values into discrete bins and then pass the discrete bin labels to stratify the data. I am not sure if we should provide support for this.

* Another method is to sort the data according to the labels and assign consecutive labels to the train and test sets according to the test ratio.

Let me know if you have any other ideas.

**Comment by Abilityguy** [comment]
Hey @rcurtin, could you take a look at this when you have the time?
I also wanted to know how we could go ahead and solve the issue raised by @zoq .

File: src/mlpack/core/data/split_data.hpp [hunk 25]
```diff
@@ -17,6 +17,136 @@
...
+  for (arma::uword i : order)
+  {
+    U label = inputLabel[i];
+    labelMap[label] += 1;
+  }
+
+  for (arma::uword labelCount : labelMap)
+  {
+      testSize += floor(labelCount * testRatio);
+      trainSize += labelCount - floor(labelCount * testRatio);
```
Comment (rcurtin) [inline src/mlpack/core/data/split_data.hpp:123]:
```suggestion
    testSize += floor(labelCount * testRatio);
    trainSize += labelCount - floor(labelCount * testRatio);
```

File: src/mlpack/core/data/split_data.hpp [hunk 26]
```diff
@@ -17,6 +17,136 @@
...
+
+  if (shuffleData)
+  {
+    order = arma::shuffle(order);
+  }
+
+  for (arma::uword i : order)
+  {
+    U label = inputLabel[i];
+    labelMap[label] += 1;
```
Comment (rcurtin) [inline src/mlpack/core/data/split_data.hpp:117]:
```suggestion
    ++labelMap[inputLabel[i]];
```
Just a simple simplification suggestion. :+1:

Comment (rcurtin) [reply src/mlpack/core/data/split_data.hpp:117]: Also, this loop is iterating in the possibly shuffled `order`, but it's not necessary to do that---and it will be faster if you access the elements of `inputLabel` in order.  (Overall it will not make a huge difference, but it's an easy change and I know it will help, so why not? :))

File: src/mlpack/core/data/split_data.hpp [hunk 27]
```diff
@@ -17,6 +17,136 @@
...
+   * Test set (1 0s, 2 1s)
+   * 0
+   * 1 1
+   */
+  size_t trainIdx = 0;
+  size_t testIdx = 0;
+  size_t trainSize = 0;
+  size_t testSize = 0;
+  arma::uvec labelMap;
+  arma::uvec testLabelMap;
```
Comment (rcurtin) [inline src/mlpack/core/data/split_data.hpp:100]: Technically these aren't maps, they're just counts, so it might be clearer to name them `labelCounts` and `testLabelCounts`.  Up to you. :+1:

**Review by rcurtin — COMMENTED** [review]
Sorry for the slightly slow response on my end.  I have just a couple comments but everything is looking good to me.  Thanks for putting time into this!  From my perspective I think it is just about ready. :+1:

File: src/mlpack/core/data/split_data.hpp (hunk 24 above)
Comment (Abilityguy) [reply src/mlpack/core/data/split_data.hpp]: Done.

**Comment by Abilityguy** [comment]
> Sorry for the slightly slow response on my end. I have just a couple comments but everything is looking good to me. Thanks for putting time into this! From my perspective I think it is just about ready.

Not an issue. :).
Thanks for the review comments!.
I have made the required documentation and style changes.

File: src/mlpack/core/data/split_data.hpp [hunk 28]
```diff
@@ -17,6 +17,138 @@
...
+  arma::uvec testLabelCounts;
+  U maxLabel = inputLabel.max();
+
+  labelCounts.zeros(maxLabel+1);
+  testLabelCounts.zeros(maxLabel+1);
+
+  arma::uvec order =
+      arma::linspace<arma::uvec>(0, input.n_cols - 1, input.n_cols);
+
+  for (arma::uword i : order)
```
Comment (rcurtin) [inline src/mlpack/core/data/split_data.hpp:112]: Sorry, just one more thing... why not iterate over `inputLabel` directly?  The compiler doesn't have any way to know that `order` is ordered, and so it won't be able to make any optimizations (at the instruction level) resulting from the linear pass over `inputLabel`.  (Of course, since the actual lookup behavior is passing through the memory in a contiguous fashion, the processor itself will manage to still load things in more quickly than a random lookup.  Anyway, still better to allow the compiler the information to know it's a linear pass.)

Comment (Abilityguy) [reply src/mlpack/core/data/split_data.hpp]: Right, that makes sense. I have made the change.

**Comment by Abilityguy** [comment]
Oops. looks I missed out testing the stratify_data flag in mlpack/tests/main_tests/preprocess_split_test.cpp. 
I will add a few tests in there.

**Comment by Abilityguy** [comment]
I have added tests in ```preprocess_split_test.cpp``` to check the bindings. 
This PR should be ready for merge now.

Also, I noticed that data was loaded using the below snippet in ```preprocess_split_test.cpp```.
```
data::Load("vc2.csv", inputData);
```
If the file doesn't exist or there is an error in loading the file, there are instances where the test case can pass (if test case depends on input size) even though there is an error in loading the file.

A better way to load data would be 
```
  if (!data::Load("vc2.csv", inputData))
    FAIL("Cannot load train dataset vc2.csv!");
```

In this case, if the file is not loaded properly, the error message comes up in the testing log. 
A quick search through the catch testing files showed most files follow this convention in the testing suite, though there are a few files where this is not followed. Some of these are

* nbc_test.cpp
* feedforward_network_test.cpp
* krann_search_test.cpp
* random_forest_test.cpp
* facilities_test.cpp

Even if the test doesn't depend on the input size in any way, I think it would be a good idea to add the ```FAIL()``` message just to ensure better testing and debugging.

I could open an issue on this and I think this could be a good first issue to introduce new contributors to the codebase.
What are your thoughts on this?

**Review by rcurtin — APPROVED** [review]
Looks good, thanks for taking the time to implement and tune this! :+1:  Do you want to add a note to `HISTORY.md` too?

**Comment by rcurtin** [comment]
> I could open an issue on this and I think this could be a good first issue to introduce new contributors to the codebase.
 What are your thoughts on this?

Totally agreed---if you'd like to take the lead on that, it would be great! :+1:

**Comment by Abilityguy** [comment]
> Looks good, thanks for taking the time to implement and tune this!  Do you want to add a note to `HISTORY.md` too?

Yup. Added a note to ```HISTORY.md```. Thanks for the detailed reviews. I learned a lot while coding this feature! :) 

> Totally agreed---if you'd like to take the lead on that, it would be great! 👍🏼

Sure. I will open up an issue soon.

**Comment by rcurtin** [comment]
Thanks @Abilityguy!
